Fix too restrictive network interface name checking

TheDiveO opened the pull request:
Fix too restrictive network interface name checking
  
`parsePodNetworkObjectName()` had too restrictive network interface name checking, trying to enforce RFC 1123 DNS label syntax. However, https://github.com/K8sNetworkPlumbingWG/multi-net-spec/blob/master/%5Bv1%5D%20Kubernetes%20Network%20Custom%20Resource%20Definition%20De-facto%20Standard.md states:

> 4.1.2.1.5 "interface"
>This optional key with value of type string requires the implementation to use the given name for the pod interface resulting from this network attachment. This keys value must be a valid Linux kernel interface name

Fixes #213 and adds tests for correct parsing behavior.

TheDiveO commented on k8sclient/k8sclient.go:
```diff
@@ -151,34 +151,42 @@ func getPodNetworkAnnotation(client KubeClient, k8sArgs *types.K8sArgs) (string,
-	allItems := []string{netNsName, networkName, netIfName}
+	allItems := []string{netNsName, networkName}
 	for i := range allItems {
 		matched, _ := regexp.MatchString("^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", allItems[i])
```
That's only _part_ of the picture, as @ and / will become valid, yet break splitting. So the splitting algorithm needs to be changed to take care of the cases that are now valid. The associated test cases take care of ensuring that the changed algo doesn't break existing things, yet covers the new cases.

dougbtv commented:
Also some examples in practice: https://paste.fedoraproject.org/paste/Cm2556XZYIO9IPbZaIWDoQ

TheDiveO commented:
In the light of `dev_valid_name` I will reconsider my code to adhere to the nif name restrictions implemented in the Linux kernel.

rkamudhan commented:
@TheDiveO As a follow up, you planning to take the @dcbw and @dougbtv suggestion in PR ? if so, we can move this PR as WIP

TheDiveO commented:
Yes, I do, so please move to WIP, as I want to bring my PR in sync with the better knowledge about what valid interfaces names in Linux are.

TheDiveO commented:
@rkamudhan I've updated my PR to now take `dev_valid_name` rules into account; I've reverted to the original separation sequence to first split on "/" and then "@", with the latter separation process now accepting multiple "@"s and separating only at the first "@". I've also updated tests too.

For the tests I noticed that it seems to be acceptable to have an empty namespace and network name, and only a network interface name, such as in "@eth42": is this correct?

TheDiveO commented:
@rkamudhan @dcbw @dougbtv any interest in merging this after I've implemented the suggestions raised before?

s1061123 commented:
let me close it because RFC1123 is not same as linux kernel interface name convention.
